src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,642 @@
+                    mem.bytesAsSlice(u64, buf[offsets.literals..offsets.end])),
+            };
+        }
+
+        fn encode(self: Self, dst: []u8, bitset: DynamicBitSetUnmanaged) void {
+            for (dst) |*b| b.* = 0;
+
+            const slices = self.parse(dst);
+            slices.total_blocks.* = @intCast(u32, bitset.bit_length);
+            slices.run_or_literal_count.* = @intCast(u32, self.runs + self.literals);
+            const word_count = bitsetMasksLen(bitset);
+
+            var run_or_literal_bit: usize = 0;
+            var run_index: usize = 0;
+            var literals_index: usize = 0;
+            var word_index: usize = 0;
+            while (word_index < word_count) {
+                const word = bitset.masks[word_index];
+                if (word == 0 or word == ~@as(usize, 0)) {
+                    const run_len = 1 + bitsetRunLength(bitset, word_index + 1, word_count, word);
+                    slices.run_lengths[run_index] = @intCast(u8, run_len - 1);
+                    if (word != 0) setBit(slices.run_of_zeroes_or_ones, run_index);
+                    setBit(slices.run_or_literal, run_or_literal_bit);
+                    run_index += 1;
+                    word_index += run_len;
+                } else {
+                    slices.literals[literals_index] = word;
+                    literals_index += 1;
+                    word_index += 1;
+                }
+                run_or_literal_bit += 1;
+            }
+
+            assert(word_index == word_count);
+            assert(slices.run_or_literal_count.* == run_or_literal_bit);
+            assert(slices.run_lengths.len == run_index);
+            assert(slices.literals.len == literals_index);
+        }
+
+        fn decode(data: []const u8, allocator: *mem.Allocator) !DynamicBitSetUnmanaged {
+            const run_or_literal_count = mem.readIntSliceLittle(u32, data[4..8]);
+            const run_or_literal_words = divCeil(usize, run_or_literal_count, 64);
+            const run_or_literal = mem.bytesAsSlice(u64, data[8 .. 8 + 8 * run_or_literal_words]);
+            var runs: usize = 0;
+            for (run_or_literal) |rol| runs += @popCount(u64, rol);
+            const decoder = BitsetEncoder(.immutable).init(.{
+                .runs = runs,
+                .literals = run_or_literal_count - runs,
+            });
+            const slices = decoder.parse(data);
+            assert(slices.run_or_literal_count.* == run_or_literal_count);
+
+            var bitset = try DynamicBitSetUnmanaged.initFull(slices.total_blocks.*, allocator);
+            errdefer bitset.deinit(allocator);
+
+            var run_index: usize = 0;
+            var literals_index: usize = 0;
+            var blocks_index: usize = 0;
+            for (mem.bytesAsSlice(u64, slices.run_or_literal)) |rol_word, rol_i| {
+                const last = rol_i == run_or_literal_words - 1;
+                if (!last and rol_word == ~@as(u64, 0)) {
```

Comment:
In fact, see also my other comment, where we can eliminate the padding case entirely.

Reply:
> In fact, see also my other comment, where we can eliminate the padding case entirely.

What comment are you referring to? I don't see a way around handling the last padding case.

Reply:
The one re: making the API stricter so that total blocks (`block_count`) must always be a perfect multiple of shard size (with shard size coincidentally also set to `8 * cache_line` to be a power of two for fast division). We should then never have any padding anywhere. 

e.g. Actual total addressable blocks when setting up a cluster must always be a multiple of 32 MiB (`65536B block size * 8 * 64B`). This is enforced at the API layer for BlockFreeSet.

Reply:
Hope that makes sense?

Reply:
A `block_count` that is a multiple of the `shard_size` doesn't necessarily mean that the last word of `run_or_literal` is unpadded.

For example, consider the case where `blocks_count = shard_size` and every block is free. That is a run of `blocks_count / bits_per_word = 5120 / 64 = 80` words. Since that is less than the maximum words/run of 256, it is compressed to a single run (`run_or_literal_count = 1`). The last (and in this case the only) word of `run_or_literal` must be padded with 63 additional bits.

Reply:
Ah! Of course. Thanks for explaining so patiently to me.

I think there's also no way either to have a minimum count for `blocks_count` either (e.g. multiply `shard_size` by a further factor, and make this the new multiple, in the hopes of reaching an LCD that always works out) since runs are variable-length.

Okay, let's still shorten the hot loop to exclude doing the branch for the last word every iteration of the hot loop, and then do the last word branch outside the hot loop. This would still make sense, right?

Reply:
Yes, that should work. Thanks!